### DISCUSSION TITLE
Embedded SmartPort support (SofSP-style)...

### DISCUSSION THREAD

**Discussion by AncientBits** [markadev/AppleII-VGA#54] [Ideas]
Hello,everyone!

With the advent of FujiNet for Apple II platform, I wondered if it is possible to add or include (existing) ROM-code on A2-VGA Pico to enable SmartPort booting (similar to SoftSP) needed for FujiNET and without having to sacrifice another precious slot...

Would this be possible? This would be REALLY nice!!

Thanks!

**Comment by markadev** [comment]
Hey there, I think it is technically possible to add a simulated ROM to the VGA card - with some caveats...

The first one is that the VGA board does not have access to the slot's IOSEL signal, which is the signal used to address the slot's ROM contents.  The VGA firmware does have access to the full address of every bus cycle but it wouldn't know what slot it was installed in to be able to respond to the correct address range for the slot.  An implementation could, for example, always place this simulated ROM in slot 5's ROM address space but it would require that any real card in slot 5 has no expansion ROM so it just kinda moves the problem.

Another challenge would be to add support to the firmware to be able to respond to bus reads with the ROM data.  I have some prototype code that could do this (to host a simulated Videx ROM) but was unable to get it working reliably at the stock Pico speeds.  Overclocking the Pico is an option but that could create other hard-to-debug stability issues so I try to avoid that.

I think multifunction cards are a good idea but the current VGA card hardware is pretty limiting.  Now that I think about it, a multifunction serial/parallel/serial-wifi/clock/virtual ROM card (like a Mountain CPS multifunction card on steroids) would be nice to have and would probably fit into one Pico.

**Reply by AncientBits** [reply]
A BIG thanks for your reply!

Well, in all fairness, this should be simple: insert A2VGA in any slot, and have extra option (or register or bits) in config menu to manually set the "slot" where the SmartPort ROM itself should be appear... and that's about it, nothing else.

The idea is for the Aii/e boot process to find this rom (and its dispatcher) to boot any SmartPort directed by such rom. This will immediately enable FujiNet, via DB19 Disk-II adapter, for instance.

Don't know how it would be (or not) handled from a bus-signaling point of view, though... but that should be sufficient!

Cheers!

**Reply by markadev** [reply]
Agreed that's a good idea for handling the virtual ROM location.  The Mountain CPS multifunction card I mentioned does a similar configuration trick for its "phantom slots" for the multiple functions.
